src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+                begin = rvalue;
+                end = strrchr(rvalue, ':');
+                if (!end || !*(end + 1)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0, "Unable to find port of endpoint: %s", rvalue);
+                        return 0;
+                }
+                len = end - begin;
+                ++end;
+        }
+
+        host = strndup(begin, len);
+        if (!host)
+                return log_oom();
+
+        port = strdup(end);
+        if (!port)
+                return log_oom();
+
+        endpoint->peer = peer;
+        endpoint->host = host;
+        endpoint->port = port;
+        LIST_PREPEND(endpoints, w->unresolved_endpoints, endpoint);
+
+        peer = NULL;
+        host = NULL;
+        port = NULL;
+        endpoint = NULL;
+
+        return 0;
+}
+
+int config_parse_wireguard_keepalive(const char *unit,
+                                     const char *filename,
+                                     unsigned line,
+                                     const char *section,
+                                     unsigned section_line,
+                                     const char *lvalue,
+                                     int ltype,
+                                     const char *rvalue,
+                                     void *data,
+                                     void *userdata) {
+        int r;
+        uint16_t keepalive = 0;
+        Wireguard *w;
+        WireguardPeer *peer;
+
+        assert(rvalue);
+        assert(data);
+
+        w = WIREGUARD(data);
+
+        assert(w);
+
+        peer = wireguard_peer_new(w, section_line);
+        if (!peer)
+                return log_oom();
+
+        if (streq(rvalue, "off"))
+                keepalive = 0;
+        else {
+                r = safe_atou16(rvalue, &keepalive);
+                if (r < 0)
+                        log_syntax(unit, LOG_ERR, filename, line, r, "The persistent keepalive interval must be 0-65535. Ignore assignment: %s", rvalue);
+        }
+
+        peer->persistent_keepalive_interval = keepalive;
+        return 0;
+}
+
+static void wireguard_init(NetDev *netdev) {
+        Wireguard *w;
+
+        assert(netdev);
+
+        w = WIREGUARD(netdev);
+
+        assert(w);
+
+        w->flags = WGDEVICE_F_REPLACE_PEERS;
+}
+
+static void wireguard_done(NetDev *netdev) {
+        Wireguard *w;
+        WireguardPeer *peer, *peer_next;
+        WireguardIpmask *mask, *mask_next;
+
+        assert(netdev);
+
+        w = WIREGUARD(netdev);
+        assert(!w->unresolved_endpoints);
+
+        w->resolve_query = sd_resolve_query_unref(w->resolve_query);
+        w->resolve_retry_event_source = sd_event_source_unref(w->resolve_retry_event_source);
+
+        LIST_FOREACH_SAFE(peers, peer, peer_next, w->peers) {
+                LIST_FOREACH_SAFE(ipmasks, mask, mask_next, peer->ipmasks) {
+                        free(mask);
+                }
+                free(peer);
```

Comment:
hmpff... i don't like this, you leave the list basically intact here, linked and everything, just free all the memory... please do this instead:

```c
while ((m = peer->ipmasks)) {
        LIST_REMOVE(…, peer->ipmasks, m);
        free(m);
}
```

And similar for the out list.

That said, maybe both objects should instead get proper destructor calls that unregister them from the containing object